STYLE.md

```diff
@@ -0,0 +1,333 @@
+
+> ```csharp
+> public class ArrayList : ICloneable, System.Collections.IList
+> ```
+
+If necessary, this should be followed by one or more paragraphs of more in-depth description.
+
+#### Examples
+
+Next, within a subheader named "Examples," include a code block with examples. 
+
+To the extent possible, this code block should be freestanding. If it depends on external definitions that aren't included in the standard library or are not otherwise automatically accessible, add a note so that users know what they need to supply themselves (e.g. that the `mod module_bindings;` refers to the `quickstart-chat` module). Do not be afraid to paste the same "header" or "prelude" code (e.g. a table declaration) into a whole bunch of code blocks, but try to avoid making easy-to-miss minor edits to such "header" code.
+
+Add comments to this code block which describe what it does. In particular, if the example prints to the console, show the expected output in a comment. [For example](https://learn.microsoft.com/en-us/dotnet/api/system.collections.arraylist?view=net-8.0):
+
+> ```csharp
+> using System;
+> using System.Collections;
+> public class SamplesArrayList  {
+>
+>    public static void Main()  {
+>
+>       // Creates and initializes a new ArrayList.
+>       ArrayList myAL = new ArrayList();
+>       myAL.Add("Hello");
+>       myAL.Add("World");
+>       myAL.Add("!");
+>
+>       // Displays the properties and values of the ArrayList.
+>       Console.WriteLine( "myAL" );
+>       Console.WriteLine( "    Count:    {0}", myAL.Count );
+>       Console.WriteLine( "    Capacity: {0}", myAL.Capacity );
+>       Console.Write( "    Values:" );
+>       PrintValues( myAL );
+>    }
+>
+>    public static void PrintValues( IEnumerable myList )  {
+>       foreach ( Object obj in myList )
+>          Console.Write( "   {0}", obj );
+>       Console.WriteLine();
+>    }
+> }
+>
+>
+> /*
+> This code produces output similar to the following:
+>
+> myAL
+>     Count:    3
+>     Capacity: 4
+>     Values:   Hello   World   !
+> 
+> */
+> ```
+
+#### Child items
+
+If the described item has any children (e.g. properties and methods of classes, variants of enums), include one or more tables for those children, as described above, followed by subsections for each child item. These subsections follow the same format as for the parent items, with a header, declaration, description, examples and tables of any (grand-)children.
+
+If a documentation page ends up with more than 3 layers of nested items, split it so that each top-level item has its own page.
+
+### Grammars and syntax
+
+TODO: Figure out the best way to format grammars. What's in the SQL ref looks good. Find out how we made the flow chart graphics, and encourage using those.
```

Comment:
I believe Mario made them, however Josh opted to go with a text based ebnf for SQL in the future: https://github.com/clockworklabs/SpacetimeDBPrivate/edit/master/proposals/0011-sql-api.md

Which tbh I find more readable and simplier.

Reply:
I don't love EBNF for broad-appeal reference documentation. I don't hate it either, but I think a lot of our users will look at it, not know what it is or how to read it, and give up on reading the docs. I don't particularly think that the flow charts are better by this metric, though. The EBNF at least has the advantage that it's unambiguous and familiar to a certain subset of our users. Perhaps what we want is to use the EBNF, but then to also include a whole lot of examples so that people who aren't used to thinking about CFGs, but are able to read SQL snippets.